src/ServiceControl/Monitoring/EndpointInstanceMonitoring.cs

```diff
@@ -75,6 +69,30 @@ public void DetectEndpointFromPersistentStore(EndpointDetails endpointDetails, b
             endpoints.GetOrAdd(endpointInstanceId.UniqueId, id => new EndpointInstanceMonitor(endpointInstanceId, monitored, domainEvents));
         }
 
+        public async Task DetectEndpointsFromBulkIngestion(IEnumerable<EndpointDetails> observedEndpoints)
+        {
+            var newEndpoints = new Lazy<List<EndpointDetails>>();
```

Comment:
The lazy is not really helping here. In case you would like to save the list allocation, if would be better to only initialize the collection when you need it. 

Here are some guidelines from my talk

- Use `Array.Empty<T>` to represent empty arrays
- Use `Enumerable.Empty<T>` to represent empty enumerables
- When the size of the items to be added to the collection are known upfront, initialize the collection with the correct count to prevent the collection from growing and thus allocating more and reshuffling things internally
- Use the concrete collection type instead of interfaces or abstract types. For example, when enumerating through a `List<T>` with `foreach` it uses a non-allocating `List<T>.Enumerator` struct. But when it is used through for example `IEnumerable<T>` that struct is boxed to `IEnumerator<T>` in the foreach and thus allocates. Of course for public APIs it is sometimes better to use the most generic type to be as broadly applicable as possible. Mileage may vary.
- With more modern CSharp versions that have good pattern matching support, it is sometimes possible to do a quick type check and based on the underlying collection type get access to the count without having to use `Count()`. With .NET 6, and later, it is also possible to use [`Enumerable.TryGetNonEnumeratedCount`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.trygetnonenumeratedcount) which internally does collection type checking to get the count without enumerating.
- Wait with instantiating collections until you really need them.

This thing is a bit of a tricky beast though. Because you are exposing on the even an array but because you don't  know how many endpoints you really observed you need to track them in a list and then expose it as an array. So the best option would be in case you want to safe allocations to forget the nice things about "immutability", highest collection type possible etc. and then do the following:

- Instead of IEnumerable<EndpointDetails> pass in at least `IReadOnlyCollection` because that exposes a count
- Lazy initialize the `List<EndpointDetails>` with the count when you need it to prevent it from growing at runtime. Another option is to optimize for the path that you mostly don't get new endpoints or just one (if that is true) and then work with a local variable or a List of one item
- Change the domain event type to be List or `IReadOnlyList` to avoid copying again